javaslang/src/main/java/javaslang/collection/AbstractsQueue.java

```diff
@@ -0,0 +1,181 @@
+        Objects.requireNonNull(elements, "elements is null");
+
+        return List.ofAll(elements).foldLeft((Q) this, AbstractsQueue<T, Q>::enqueue); // TODO crashes the compiler for Q::enqueue
+    }
+
+    /**
+     * Returns the first element without modifying it.
+     *
+     * @return the first element
+     * @throws NoSuchElementException if this Queue is empty
+     */
+    public T peek() {
+        if (isEmpty()) {
+            throw new NoSuchElementException("peek of empty " + getClass().getSimpleName());
+        } else {
+            return head();
+        }
+    }
+
+    /**
+     * Returns the first element without modifying the Queue.
+     *
+     * @return {@code None} if this Queue is empty, otherwise a {@code Some} containing the first element
+     */
+    Option<T> peekOption() {
+        return isEmpty() ? Option.none() : Option.some(peek());
+    }
+
+    /**
+     * Dual of {@linkplain #tail()}, returning all elements except the last.
+     *
+     * @return a new instance containing all elements except the last.
+     * @throws UnsupportedOperationException if this is empty
+     */
+    @Override
+    public abstract Q init();
+
+    /**
+     * Dual of {@linkplain #tailOption()}, returning all elements except the last as {@code Option}.
+     *
+     * @return {@code Some(Q)} or {@code None} if this is empty.
+     */
+    @Override
+    public Option<Q> initOption() {
+        return isEmpty() ? Option.none() : Option.some(init());
+    }
+
+    /**
+     * Drops the first element of a non-empty Traversable.
+     *
+     * @return A new instance of Traversable containing all elements except the first.
+     * @throws UnsupportedOperationException if this is empty
+     */
+    @Override
+    public abstract Q tail();
+
+    @Override
+    public Option<Q> tailOption() {
+        return isEmpty() ? Option.none() : Option.some(tail());
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Q retainAll(Iterable<? extends T> elements) {
+        return Collections.retainAll((Q) this, elements);
+    }
+
+    @SuppressWarnings("unchecked")
+    public Q removeAll(Iterable<? extends T> elements) {
+        return Collections.removeAll((Q) this, elements);
+    }
+
+    @Override
+    public Q takeWhile(Predicate<? super T> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        return takeUntil(predicate.negate());
+    }
+
+    @Override
+    public abstract Q takeUntil(Predicate<? super T> predicate);
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public Q peek(Consumer<? super T> action) {
+        Objects.requireNonNull(action, "action is null");
+        if (!isEmpty()) {
+            action.accept(head());
+        }
+        return (Q) this;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        return o == this || o instanceof Iterable && Collections.equals(this, (Iterable) o);
```

Comment:
checking for type `Iterable` does not work. `equals` needs to be symmetric: if `this` is an `AbstractQueue(1,2,3)` and `that` is an `Iterable(1,2,3)` then the current impl says that `this.equals(that)` is true. But `that.equals(this)` is not true when that checks `instanceof` a specific type that is not an `AbstractQueue`.

Example: AbstractMultimap.equals:

```java
@Override
    public boolean equals(Object o) {
        if (o == this) {
            return true;
        } else if (o != null && getClass().isAssignableFrom(o.getClass())) {
            final AbstractMultimap<?, ?, ?> that = (AbstractMultimap<?, ?, ?>) o;
            return this.back.equals(that.back);
        } else {
            return false;
        }
    }
```